### PR TITLE
Adds basic admin functions: Crud Users and Groups

### PR DESCRIPTION
**Pull request by clemensmanert**
Replaces user name by mail for user identificaiton.

### CODE REVIEW

File: routes/admin.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            return redirect("/admin/login", code=403)
+
+        if target_user.role != RoleType.ADMINISTRATOR:
+            return redirect("/admin/login", code=403)
+        return fn(*args, **kwargs)
+    return wrapper
+
+
+@Admin.route('/favicon.ico', methods=['GET'])
+def favicon():
```
Comment (missgreenwood) [inline routes/admin.py:35]: Very good! :)

File: routes/admin.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    """Prevents the logger from logging 404 when the browser asks for a favicon.
+    """
+    return ''
+
+@Admin.route('/admin/login', methods=['POST'])
+def do_admin_login():
+    """Logs in the given user.
```
Comment (missgreenwood) [inline routes/admin.py]: Suggestion: "Log-in for the given user."

File: routes/admin.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+@Admin.route('/admin/login')
+def admin_login():
+    """Returns a login form.
+    """
+    return Response(render_template('admin/login.html'))
+
+@Admin.route('/admin/logout')
+def logout():
+    """Performs a logout on the current user.
```
Comment (missgreenwood) [inline routes/admin.py]: "Performs a logout for the current user."

File: routes/admin.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    """
+    resp = Response(render_template('admin/login.html',
+                                    message='Your session has been destroyed.'))
```
Comment (missgreenwood) [inline routes/admin.py]: Better: "Your session has been canceled."

File: routes/admin.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+def admin():
+    """Returns the admin home page.
+    """
+    return Response(render_template('admin.html'))
+
+
+@Admin.route('/admin/user/create', methods=['POST'])
+@admin_required
+def do_user_create():
+    """Creates a new user. Generats a password token. Does not send a mail.
```
Comment (missgreenwood) [inline routes/admin.py]: "Generates"

File: routes/admin.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    :param str gender: The new user's gender.
+    :param str name: The new user's name.
+    :param str mail:The new user's mail.
```
Comment (missgreenwood) [inline routes/admin.py]: ":param str mail: The new user's mail."

File: routes/admin.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    :param str group_id: The new user's group id.
+    """
+    target = User(
+        request.form['gender'],
+        request.form['name'],
+        request.form['mail'],
+        request.form['meter_id'],
+        request.form['group_id'],
+        secrets.token_hex(33))
+    target.set_role(request.form['role'])
```
Comment (missgreenwood) [inline routes/admin.py:109]: Please add the parameter 'role' to the function parameters.

File: routes/admin.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    db.session.commit()
+    return user_list("Created user " + target.get_name())
+
+@Admin.route('/admin/user/create', methods=['GET'])
+@admin_required
+def request_user_create():
+    """Returns a form to create a new user.
+    """
+    return Response(render_template('admin/user/create-update.html',
+                                    csrf_token=(get_raw_jwt() or {}).get("csrf"),
```
Comment (missgreenwood) [inline routes/admin.py:122]: Please explain this line a little bit.

File: routes/admin.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+                                    target="/admin/user/create",
+                                    genders=list(GenderType),
+                                    states=list(StateType),
+                                    groups=Group.query.all(),
+                                    roles=list(RoleType),
+                                    gender=GenderType.FEMALE,
```
Comment (missgreenwood) [inline routes/admin.py:128]: How so?

File: routes/admin.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+                                    meter_id=target_user.meter_id,
+                                    group_id=target_user.group_id,
+                                    role=target_user.role,
+                                    state=target_user.state),
+                    mimetype='text/html')
+
+@Admin.route('/admin/user/delete', methods=['GET'])
+@admin_required
+def user_delete():
+    """Deletes an user.
```
Comment (missgreenwood) [inline routes/admin.py]: "Deletes a user."

File: routes/admin.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    :param str id: The id to identify the user to update.
```
Comment (missgreenwood) [inline routes/admin.py]: "delete" instead of "update"

File: routes/admin.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+    return Response(render_template('admin/user/list.html',
+                                    users=User.query.all(),
+                                    message=message),
+                    mimetype='text/html')
+
+@Admin.route('/admin/group/create', methods=['POST'])
+@admin_required
+def do_group_create():
+    """Creates a group.
+    :param str name: The new groups name.
```
Comment (missgreenwood) [inline routes/admin.py]: "The new group's name."

File: routes/admin.py [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    :param str meter: The new groups meter.
```
Comment (missgreenwood) [inline routes/admin.py]: Better: "The new group's meter id."

File: routes/admin.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+                                    csrf_token=(get_raw_jwt() or {}).get("csrf"),
+                                    target="/admin/group/create"),
+                    mimetype='text/html')
+
+@Admin.route('/admin/group/update', methods=['POST'])
+@admin_required
+def do_group_update():
+    """Updates an existing group.
+    :param str id: The id of the group to update.
+    :param str name: The groups name.
```
Comment (missgreenwood) [inline routes/admin.py:242]: "The group's name."

File: routes/admin.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+    :param str meter: The groups meter.
```
Comment (missgreenwood) [inline routes/admin.py:243]: "The group's meter id."

File: routes/login.py [hunk 16]
```diff
@@ -19,20 +20,22 @@ def login():
       404: If the useraccount is not active.
     """
     j = request.get_json(force=True)
-    user_requested = j['user']
+    user_requested = j['user'].lower()
```
Comment (missgreenwood) [inline routes/login.py:23]: Do we really want this?

File: routes/login.py [hunk 17]
```diff
@@ -19,20 +20,22 @@ def login():
...
 
-    targetUser = User.query.filter_by(name=user_requested).first()
-    if targetUser is None:
-        return Error('Unknown credentials',
-                     'Try again with proper username/password.').to_json(), 401
+    target_user = User.query.filter_by(mail=user_requested).first()
+    if target_user is None:
+        return (Error('Unknown credentials',
+                      'Try again with proper username/password.').to_json(),
+                status.HTTP_401_UNAUTHORIZED)
```
Comment (missgreenwood) [inline routes/login.py:30]: Very good! :+1:

File: routes/reset_password.py [hunk 18]
```diff
@@ -25,7 +25,7 @@ def request_password_reset_token():
     :param str user: The user's mail address.
     """
     j = request.get_json(force=True)
-    user_requested = j['user']
+    user_requested = j['user'].lower()
```
Comment (missgreenwood) [inline routes/reset_password.py:28]: Cf. above.

File: routes/set_password.py [hunk 19]
```diff
@@ -29,11 +29,11 @@ def set_password():
         404: If the user is not found.
     """
     j = request.get_json(force=True)
-    user_requested = j['user']
+    user_requested = j['user'].lower()
```
Comment (missgreenwood) [inline routes/set_password.py:32]: Cf. above.

File: setup_app.py [hunk 20]
```diff
@@ -22,6 +24,14 @@ def setup_app(app_config):
     """
     app = Flask(__name__)
     app.config.from_object(app_config)
+    # Look for jwt token in the headers and the cookies, the headers are used
+    # by the app; the cookies by the admin ui.
```
Comment (missgreenwood) [inline setup_app.py]: "UI"

**Review by missgreenwood — CHANGES_REQUESTED** [review]
As we discussed, please fix the build errors before merging this PR. Otherwise, everything looks fine to me (haven't looked at the html files in detail, though ;)).

File: routes/admin.py (hunk 9 above)
Comment (clemensmanert) [reply routes/admin.py:128]: Pre selected value. I've added a comment.

File: routes/admin.py (hunk 10 above)
Comment (clemensmanert) [reply routes/admin.py]: whops again

File: routes/login.py (hunk 16 above)
Comment (clemensmanert) [reply routes/login.py:23]: ja, cause it usually a mail address.

**Pull request merged**